javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -0,0 +1,412 @@
+/**
+ * Provides classes for working with [Spife](https://github.com/npm/spife) applications.
+ */
+
+import javascript
+import semmle.javascript.frameworks.HTTP
+
+/**
+ * Provides classes for working with [Spife](https://github.com/npm/spife) applications.
+ */
+module Spife {
+  /**
+   * A call to a Spife method that sets up a route.
+   */
+  private class RouteSetup extends API::CallNode, Http::Servers::StandardRouteSetup {
+    TaggedTemplateExpr template;
+
+    RouteSetup() {
+      this.getCalleeNode().asExpr() = template and
+      API::moduleImport(["@npm/spife/routing", "spife/routing"])
+          .asSource()
+          .flowsToExpr(template.getTag())
+    }
+
+    private string getRoutePattern() {
+      // Concatenate the constant parts of the expression
+      result =
+        concat(Expr e, int i |
+          e = template.getTemplate().getElement(i) and exists(e.getStringValue())
+        |
+          e.getStringValue() order by i
+        )
+    }
+
+    private string getARouteLine() {
+      result = this.getRoutePattern().splitAt("\n").regexpReplaceAll(" +", " ").trim()
+    }
+
+    private predicate hasLine(string method, string path, string handlerName) {
+      exists(string line | line = this.getARouteLine() |
+        line.splitAt(" ", 0) = method and
+        line.splitAt(" ", 1) = path and
+        line.splitAt(" ", 2) = handlerName
+      )
+    }
+
+    API::Node getHandlerByName(string name) { result = this.getParameter(0).getMember(name) }
```

Comment:
You also have to remove the rest of the uses of `API::Node` in this class.

Reply:
@erik-krogh how would you rewrite this predicate? I thought about this `result = this.getACallee().getParameter(0).(DataFlow::ObjectLiteralNode).getAPropertyRead(name)` but ObjectLiteralNode is not compatible with Parameter

Reply:
Just remove `.(DataFlow::ObjectLiteralNode)` from your attempt and you should be good.  
`getAPropertyRead()` is defined on `SourceNode`, and both `ObjectLiteralNode` and `ParameterNode` are `SourceNode`s.

Reply:
Ok, I was not getting that option from the autocomplete menu, but I was missing a wrapping call to `DataFlow::parameterNode()`